neps/nep-0605-sharded-contracts.md

```diff
@@ -400,28 +352,171 @@ to store the sharded contract code on the shard.
+    /// information about which sharded contract to use
+    sharded_contract: ShardedContractType,
+}
+```
 
-By creating a new account, we get the necessary isolation that is important to ensure that no malicious or accidental tampering can happen.  In particular, we prevent the following potential attacks and possibly more:
+This action is similar to the `CreateAccountAction` action.  Further, this action can be called multiple times to use multiple different sharded contract codes on the same account.  To support this action, we propose that each time this action is issued, a new subordinate account is created.  Details of subordinate accounts are discussed below.
 
-1. If an account has a contract code `C` deployed on it and it is using a sharded contract code `S`, it might be possible for them to access each other's state and corrupt it.
-2. Similarly, if an account is using multiple sharded contract codes `S1` and `S2`, then it is possible that they could access and corrupt each other's state.
-3. Sharded contract codes have unfettered access to the NEAR tokens on the account.  They could transfer it away or they could keep increasing the size of the state increasing the amount of NEAR tokens that are locked up.
+#### `ShardedFunctionCallAction`
 
-By creating a separate subordinate account for each sharded contract that an account wants to use, we ensure that the account is able to specify limits on how much resources the sharded contract can use.  We also ensure that the sharded contract is well isolated from the account and from other sharded contracts which might corrupt its state.
+Once an account is using a sharded contract code, the sharded contract code can be called using this action.  This action is similar to the `FunctionCallAction` action.  It contains additional information about what type of sharded contract to call on the account.  Note that as seen in the FT example above, this action will also be used when a sharded contract code calls another one.
 
-XXX: Discuss the design further.  Some open questions
+```rust
+/// This is similar to the existing `FunctionCallAction`.  `FunctionCallAction`
+/// allows calling contract codes that are deployed using the
+/// `DeployContractAction` or the `UseGlobalContractAction` on an account.  This
+/// action allows calling contract codes that are deployed using the
+/// `UseShardedContractAction`.
+struct ShardedFunctionCallAction {
+    // An account can be using multiple sharded contract codes.  This identifies
+    // which one should be called.
+    receiver_sharded_contract: ShardedContractType,
+    // An account can be using multiple sharded contract codes.  This identifies
+    // which contract code on the predecessor is sending the action.  If this is
+    // `None`, then the caller is not using sharded contract code.
+    predecessor_sharded_contract: Option<ShardedContractType>,
+    // additionally arguments are identical to `FunctionCallAction`.
+}
```

Comment:
Should this include min/max version fields?

Reply:
Or should the receiver be of type `CallShardedContractReceiver`?

Reply:
yes, you are right, updating to `CallShardedContractReceiver`.